compatibility.bs

```diff
@@ -832,6 +847,129 @@ supported on all {{Window}} objects and <code><{body}></code> elements as attrib
+
+The <dfn>redirect heuristic grant duration</dfn> is an [=implementation-defined=]
+[=duration=] that represents the length of time after the redirect heuristic is
+triggered that a user agent will store the corresponding redirect heuristic grant
+in its [=heuristic grants map=].
+
+Note: Most implementations use 15 minutes as the [=redirect heuristic grant duration=].
+
+<h3 id="sa-heuristics-access-grant">Cookie Access Grant</h3>
+
+<div algorithm>
+
+To <dfn>grant access for heuristics</dfn> given a [=site=] |site|, [=site=]
+|firstPartySite|, [=moment=] |currentWallTime|, and [=duration=] |duration|,
+perform the following steps:
+
+1. Let |key| be a [=site pair=] with [=third party site=] set to |site| and
+    [=first party site=] set to |firstPartySite|.
+1. Let |expirationTime| be |duration| after |currentWallTime|.
+1. Set [=heuristic grants map=][|key|] to |expirationTime|.
+
+</div>
+
+<h3 id="sa-heuristics-access">Cookie Access Monkey Patch</h3>
+
+Issue: TODO: add monkey patch to network fetch cookie access that reads from [=heuristic grants map=].
+
+<h3 id="sa-heuristics-popup">Popup Heuristic</h3>
+
+Append the following steps to the [[!HTML]]'s <a spec="html">activation notification</a> algorithm:
+
+<div algorithm="user-activation-patch">
+
+1. Run [=detect a popup heuristic=] given <var ignore>document</var>.
+
+</div>
+
+<div algorithm>
+
+To <dfn>detect a popup heuristic</dfn> given a [=Document=] |document|,
+perform the following steps:
+
+1. Let |browsingContext| be |document|'s [=Document/browsing context=].
+1. If |browsingContext|'s [=browsing context/is popup=] is false, then return.
+1. If |browsingContext|'s [=browsing context/opener browsing context=] is null, then return.
+1. If |browsingContext|'s [=browsing context/opener origin at creation=] is null, then return.
+1. Let |firstPartySite| be the result of running [=obtain a site=] given the |browsingContext|'s
+    [=browsing context/opener origin at creation=].
+1. Let |navigable| be |document|'s [=node navigable=].
+1. If |navigable| is null, then return.
+1. Let |topDocument| be |navigable|'s [=navigable/top-level traversable=]'s
+    [=navigable/active document=].
+1. Let |origin| be |topDocument|'s [=Document/origin=].
+1. If |origin| is an [=opaque origin=] then return.
+1. Let |site| be the result of running [=obtain a site=] given |origin|.
+1. Let |currentWallTime| be |topDocument|'s [=relevant settings object=]'s
+    [=environment settings object/current wall time=].
+1. [=Grant access for heuristics=] given |site|, |firstPartySite|, |currentWallTime|, and
+    [=popup heuristic grant duration=].
+
+Issue(amaliev/3pcd-exemption-heuristics#3): TODO: Consider whether to check for third-party iframe initiators.
+
+</div>
+
+<h3 id="sa-heuristics-redirect">Redirect Heuristic</h3>
+
+Insert the following steps in [[!HTML]]'s <a spec="html">load a document</a> algorithm, before the
```

Comment:
This probably isn't what you want. Most documents (all text/html documents among them) don't reach this step.

Reply:
Thanks for the callout, I've updated this to point to the start of the "load a document" algorithm.

Generally we prefer to call any end-of-extended-navigation algorithms after the navigation "commits" (which is a standard term in Chrome and MDN but not in the HTML spec). This is to avoid spurious calls in case a navigation aborts, e.g. on loading an invalid resource. However, the structure of the "load a document" algorithm makes it difficult to patch a step at the end of all branches, after the load completes. Anyway, in this case the effect of allowing uncommitted navigations is minor (especially due to the A-B-A requirement below).